feat(ingestion): lineage for stored procedures across SQL databases with temp table filtering

acrylJonny opened the pull request:
*Summary*

Adds *stored procedure lineage extraction* across 6 major SQL databases: *MySQL, MariaDB, PostgreSQL, Oracle, Snowflake, and SQL Server*. This enables DataHub to capture column-level lineage from stored procedure logic, including sophisticated *temporary table collapsing* to provide clean, actionable lineage graphs.

*Key Features*

*🔍 *Multi-Database Support**
- *MySQL & MariaDB*: Extracts procedures from `INFORMATION_SCHEMA.ROUTINES` with full lineage parsing
- *PostgreSQL*: Parses `pg_proc` for function/procedure definitions
- *Oracle*: Supports `ALL_PROCEDURES` and `ALL_SOURCE` views
- *Snowflake*: Integrates with `INFORMATION_SCHEMA.PROCEDURES` and access history
- *SQL Server*: Reads from `sys.procedures` and `sys.sql_modules`

*🧹 *Intelligent Temp Table Filtering**
Automatically detects and filters temporary tables from lineage graphs:
- *MySQL/MariaDB*: `CREATE TEMPORARY TABLE` patterns
- *PostgreSQL*: `CREATE TEMP TABLE` patterns  
- *Oracle*: `CREATE GLOBAL TEMPORARY` patterns
- *Snowflake*: Configurable regex patterns for ETL tool temp tables (Fivetran, DBT, etc.)

*Example*: Procedure with `A → temp_table → B` now produces clean lineage `A → B`

*📊 *Column-Level Lineage**
Extracts fine-grained column transformations:
```sql
CREATE PROCEDURE calculate_priority_customers()
BEGIN
  CREATE TEMPORARY TABLE temp_priority AS
    SELECT customer_id, total_spent FROM orders;
  
  INSERT INTO priority_customers (id, value)
    SELECT customer_id, total_spent FROM temp_priority;
END

## Checklist
Thank you for contributing to DataHub!

Before you submit your PR, please go through the checklist below:

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [x] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

acrylJonny committed:
initial commit

acrylJonny committed:
Merge branch 'master' into mysql-stored-procs

acrylJonny committed:
Update mariadb.py

acrylJonny committed:
updates for stored proc lineage

acrylJonny committed:
Merge branch 'master' into mysql-stored-procs

acrylJonny committed:
Merge branch 'datahub-project:master' into mysql-stored-procs

acrylJonny committed:
add testing

acrylJonny committed:
update to ignore certain changeable aspects

acrylJonny committed:
adding further tests for mariadb

acrylJonny committed:
Merge branch 'master' into mysql-stored-procs

acrylJonny committed:
more tests

acrylJonny committed:
updating mariadb test

acrylJonny committed:
Merge branch 'master' into mysql-stored-procs

acrylJonny committed:
Merge branch 'master' into mysql-stored-procs

acrylJonny committed:
Merge branch 'master' into mysql-stored-procs

acrylJonny committed:
test updates

acrylJonny committed:
Merge branch 'master' into mysql-stored-procs

acrylJonny committed:
Merge branch 'master' into mysql-stored-procs

acrylJonny committed:
aligning to current proc methods

acrylJonny committed:
Merge branch 'master' into mysql-stored-procs

acrylJonny committed:
updated golden mces

acrylJonny committed:
updated golden mces

acrylJonny committed:
Merge branch 'master' into mysql-stored-procs

acrylJonny committed:
Merge branch 'master' of https://github.com/acrylJonny/datahub into mysql-stored-procs

acrylJonny committed:
align to latest stored proc style

acrylJonny committed:
Merge branch 'master' into mysql-stored-procs

acrylJonny committed:
fix security vulnerabilities

acrylJonny committed:
prettier

acrylJonny committed:
Merge branch 'master' into mysql-stored-procs

acrylJonny committed:
Revert "Merge branch 'master' into mysql-stored-procs"

This reverts commit e232676267d8ca45f54c1b3345c56ebe7ee441ff, reversing
changes made to f5a58e89b640e5edaa7186d730df54443c5c604e.

acrylJonny committed:
Reapply "Merge branch 'master' into mysql-stored-procs"

This reverts commit 0a1de79c5dfccac5e88a7c866f5061674741e81e.

acrylJonny committed:
align with updates

acrylJonny committed:
temptable filter in stored procs

acrylJonny committed:
Update mysql_table_level_only.json

treff7es commented on metadata-ingestion/src/datahub/ingestion/source/sql/mysql.py:
```diff
@@ -102,6 +106,101 @@ def create(cls, config_dict, ctx):
+        MySQL temporary tables typically start with # or _tmp patterns.
+        """
+        # MySQL temporary table patterns
+        temp_patterns = [
```
I think this can be simplified a bit:
```
temp_patterns = [
	r"^#.*",  # Tables starting with #
    r"^(tmp|temp)_.*",    # Tables starting with tmp_ or temp_
    r".*_(tmp|temp)$",    # Tables ending with _tmp or _temp
    r".*_(tmp|temp)_.*",  # Tables containing _tmp_ or _temp_
]

```

acrylJonny replied:
I'll simplify this and update the commit

acrylJonny committed:
addressing comments

acrylJonny committed:
Merge branch 'master' into mysql-stored-procs

acrylJonny commented on metadata-ingestion/src/datahub/ingestion/source/sql/mysql.py:
```diff
@@ -102,6 +106,101 @@ def create(cls, config_dict, ctx):
+            r".*_temp_.*",  # Tables containing _temp_
+        ]
+
+        import re
```
moved to the top

acrylJonny commented on metadata-ingestion/src/datahub/ingestion/source/sql/mariadb.py:
I've removed this as it's not required. This will make it consistent with MySQL

acrylJonny commented on metadata-ingestion/src/datahub/ingestion/source/sql/mariadb.py:
I've improved this but left the logic as a fallback

acrylJonny commented on metadata-ingestion/src/datahub/ingestion/source/sql/mariadb.py:
```diff
@@ -7,14 +15,159 @@
+
+        return procedures_list
+
+    def _escape_identifier(self, identifier: str) -> str:
```
No longer required in MariaDB and isn't required in MySQL either

treff7es commented on metadata-ingestion/src/datahub/ingestion/source/sql/mariadb.py:
```diff
@@ -7,14 +16,134 @@
+                )
+            except Exception as e:
+                procedure_name = (
+                    routine_name if "routine_name" in locals() else "unknown"
```
What is this exactly for?

treff7es replied:
I don't get the locals() usage

acrylJonny committed:
addressing comments

acrylJonny commented on metadata-ingestion/src/datahub/ingestion/source/sql/mysql.py:
```diff
@@ -122,30 +234,39 @@ def get_procedures_for_schema(
+            procedure_rows: List[Row] = list(procedures)
             for row in procedure_rows:
+                # Extract name with fallback - name is required for BaseProcedure
+                procedure_name = getattr(row, "name", None)
```
now converted to a dict and updated in the unit tests

acrylJonny committed:
Merge branch 'master' into mysql-stored-procs

acrylJonny committed:
Merge branch 'master' into mysql-stored-procs

acrylJonny committed:
Merge branch 'master' into mysql-stored-procs

acrylJonny committed:
Merge branch 'master' into mysql-stored-procs

acrylJonny committed:
Update mysql.py

acrylJonny committed:
Merge branch 'master' into mysql-stored-procs

treff7es commented on metadata-ingestion/src/datahub/ingestion/source/sql/mysql.py:
```diff
@@ -93,21 +121,16 @@ class MySQLConnectionConfig(SQLAlchemyConnectionConfig):
+):
+    # MySQLConfig now inherits stored procedure configuration from StoredProcedureConfigMixin
+    # This includes: include_stored_procedures, procedure_pattern
+    pass
```
This shouldn't be there otherwise it just pass the rest

treff7es commented on metadata-ingestion/src/datahub/ingestion/source/sql/mariadb.py:
```diff
@@ -7,14 +15,112 @@
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+STORED_PROCEDURES_QUERY = """
```
Both MySQL and MariaDB define STORED_PROCEDURES_QUERY with slightly different column selections:
  - MySQL (mysql.py:80-93): Selects lowercase aliases (name, definition, comment)
  - MariaDB (mariadb.py:26-40): Selects uppercase aliases (ROUTINE_SCHEMA, ROUTINE_NAME, etc.)
  
  Is this correct?

treff7es commented on metadata-ingestion/src/datahub/ingestion/source/sql/mysql.py:
```diff
@@ -219,30 +332,40 @@ def get_procedures_for_schema(
-                        last_altered=None,
-                        extra_properties=None,
-                        comment=None,
+                        procedure_definition=row_dict.get("definition"),
```
- MySQL (mysql.py:357-358): Passes datetime fields directly from the query
  created=row_dict.get("CREATED"),
  last_altered=row_dict.get("LAST_ALTERED"),
  - MariaDB (mariadb.py:94-95): Parses them with _parse_datetime()
  created=self._parse_datetime(row_dict.get("CREATED")),
  last_altered=self._parse_datetime(row_dict.get("LAST_ALTERED")),

  This inconsistency is confusing. Both should use the same approach.

treff7es commented on metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema_gen.py:
```diff
@@ -176,6 +176,7 @@ def __init__(
         aggregator: Optional[SqlParsingAggregator],
         snowsight_url_builder: Optional[SnowsightUrlBuilder],
         fetch_views_from_information_schema: bool = False,
+        is_temp_table: Callable[[str], bool] = lambda _: False,
```
maybe we should call this `is_temp_table_fn` and be optional `Optional[Callable[str]]` and not `Callable[[str], bool]`

treff7es commented on metadata-ingestion/src/datahub/ingestion/source/sql/mariadb.py:
```diff
@@ -7,14 +15,112 @@
+        for row in conn.execute(query, {"schema": schema}):
+            # Convert SQLAlchemy Row to dict for easier and safer access
+            row_dict = dict(row)
+            procedure_name = row_dict.get("ROUTINE_NAME", "unknown")
```
Is set but never used in error reporting since we immediately check for None and skip

treff7es reviewed (changes requested):
I left some comment, please check

acrylJonny committed:
Merge branch 'master' into mysql-stored-procs

acrylJonny committed:
addressing comments

acrylJonny committed:
Merge branch 'master' into mysql-stored-procs

acrylJonny committed:
Merge branch 'master' into mysql-stored-procs

treff7es commented:
@acrylJonny golden files needs to be updated

acrylJonny committed:
Merge branch 'master' into mysql-stored-procs

acrylJonny committed:
adding oracle to implementation and refreshing golden files

acrylJonny committed:
Merge branch 'master' into mysql-stored-procs

acrylJonny committed:
consolidation

acrylJonny committed:
fixes

acrylJonny committed:
updating mocks

acrylJonny committed:
fixing two tier issue

acrylJonny committed:
mssql fixes for filtering

acrylJonny committed:
more reusability

acrylJonny committed:
Merge branch 'master' into mysql-stored-procs

acrylJonny commented:
Closing this PR — it has been superseded by upstream commits and split into focused follow-ups.

## Already in `master`

The core cross-database stored-procedure work has landed via separate PRs since this one was opened:

- [#14102](https://github.com/datahub-project/datahub/pull/14102) — Postgres stored procedures (merged 2025-07-22)
- [#14274](https://github.com/datahub-project/datahub/pull/14274) — MySQL stored procedures (merged 2025-08-11, hours after this PR was opened)
- [#15340](https://github.com/datahub-project/datahub/pull/15340) — MSSQL stored-procedure lineage fix
- [#16067](https://github.com/datahub-project/datahub/pull/16067) — Oracle stored procedures + PL/SQL parsing (merged 2026-02-13)

The shared `stored_procedures/` package in `master` has also been redesigned: the old `enrichment.py` module and `fetch_procedures_with_enrichment` framework this PR introduced were removed and replaced by `generate_procedure_workunits` / `generate_procedure_lineage`, so resolving the 1227-commit merge here would require rewriting most of the diff against the new API.

## Net-new content — split into smaller PRs

The pieces that aren't yet covered by the merged PRs above have been filed as focused follow-ups against current `master`:

- **#17461** — `test(mariadb): add integration test suite` (PR A: docker-compose + setup.sql + recipe + golden + `test_mariadb.py`; no production code change).
- **#17462** — `test(mysql): add integration test for stored procedure ingestion` (PR B: stored-procedure integration recipe and golden; opened as **draft** pending fresh golden regeneration against current `master`).
- **#17463** — `fix(ingest/mssql): resolve 2-part-name lineage references against discovered_datasets` (PR C: the targeted 2-part-name temp-table filtering fix originally on this branch).
- **#17464** — `feat(ingest/oracle): add include_procedure_enrichment to skip per-procedure round-trips` (PR D: the performance flag that skips the N+1 enrichment queries against `{ALL|DBA}_SOURCE`/`_ARGUMENTS`/`_DEPENDENCIES`).

Closing in favour of those four.

acrylJonny commented:
Superseded — see the comment above for the four follow-up PRs.
